File: runtime/go/stream.go
package ridlruntime

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync/atomic"
)

var ErrStreamAlreadyOpened = errors.New("an RPC stream call builder can only be opened once")

type StreamSession interface {
	Next(context.Context) (Frame, bool, error)
	Cancel(context.Context) error
}

type FramedStream interface {
	Carrier() string
	Open(context.Context, Frame) (StreamSession, error)
}

type StreamRequest struct {
	Key    string
	Method string
	Path   string
	Query  []QueryPair
	Body   json.RawMessage
}

type RpcStreamContext struct {
	ID        string
	Key       string
	Carrier   string
	Ended     bool
	Cancelled bool
	Err       error
}

type StreamRemoteError struct {
	Code    string
	Message string
}

func (e *StreamRemoteError) Error() string {
	if e.Message == "" {
		return fmt.Sprintf("remote RPC stream error %s", e.Code)
	}
	return fmt.Sprintf("remote RPC stream error %s: %s", e.Code, e.Message)
}

type RpcStreamClient[T any] struct {
	session   StreamSession
	id        string
	key       string
	carrier   string
	decode    func(json.RawMessage) (T, error)
	ended     bool
	cancelled bool
	done      bool
	err       error
}

func (c *RpcStreamClient[T]) Context() RpcStreamContext {
	return RpcStreamContext{
		ID:        c.id,
		Key:       c.key,
		Carrier:   c.carrier,
		Ended:     c.ended,
		Cancelled: c.cancelled,
		Err:       c.err,
	}
}

func (c *RpcStreamClient[T]) Cancel(ctx context.Context) error {
	if c.ended || c.cancelled {
		return nil
	}
	if err := c.session.Cancel(ctx); err != nil {
		c.err = fmt.Errorf("%s: stream cancellation failed: %w", c.key, err)
		return c.err
	}
	c.cancelled = true
	c.done = true
	return nil
}

func (c *RpcStreamClient[T]) Next(ctx context.Context) (T, bool, error) {
	var zero T
	if c.done {
		return zero, false, nil
	}
	frame, ok, err := c.session.Next(ctx)
	if err != nil {
		c.done = true
		c.err = fmt.Errorf("%s: stream transport failed: %w", c.key, err)
		return zero, false, c.err
	}
	if !ok {
		c.done = true
		if c.ended || c.cancelled {
			return zero, false, nil
		}
		c.err = errors.New("stream transport ended without an end or cancel frame")
		return zero, false, c.err
	}
	if frame.ID != c.id {
		c.done = true
		c.err = fmt.Errorf("frame for correlation id %s arrived on the stream for %s", frame.ID, c.id)
		return zero, false, c.err
	}

	switch frame.Kind {
	case Data:
		if !frame.HasBody {
			c.done = true
			c.err = errors.New("a stream data frame arrived without a body")
			return zero, false, c.err
		}
		value, err := c.decode(frame.Body)
		if err != nil {
			c.done = true
			c.err = fmt.Errorf("%s: stream data failed typed decoding: %w", c.key, err)
			return zero, false, c.err
		}
		return value, true, nil
	case End:
		c.ended = true
		c.done = true
		return zero, false, nil
	case Error:
		c.done = true
		message := ""
		if frame.Message != nil {
			message = *frame.Message
		}
		c.err = &StreamRemoteError{Code: frame.Code, Message: message}
		return zero, false, c.err
	case Cancel:
		c.cancelled = true
		c.done = true
		return zero, false, nil
	case Call:
		c.done = true
		c.err = errors.New("a call frame cannot arrive inside its response stream")
		return zero, false, c.err
	default:
		c.done = true
		c.err = fmt.Errorf("unsupported stream frame kind %q", frame.Kind)
		return zero, false, c.err
	}
}

type RpcStreamCallBuilder[T any] struct {
	stream     FramedStream
	correlator *Correlator
	request    StreamRequest
	decode     func(json.RawMessage) (T, error)
	opened     atomic.Bool
}

func NewRpcStreamCallBuilder[T any](
	stream FramedStream,
	correlator *Correlator,
	request StreamRequest,
	decode func(json.RawMessage) (T, error),
) *RpcStreamCallBuilder[T] {
	return &RpcStreamCallBuilder[T]{
		stream:     stream,
		correlator: correlator,
		request:    request,
		decode:     decode,
	}
}

// Stream is the sole I/O boundary. Constructing and configuring the builder
// cannot open the transport.
func (b *RpcStreamCallBuilder[T]) Stream(ctx context.Context) (*RpcStreamClient[T], error) {
	if !b.opened.CompareAndSwap(false, true) {
		return nil, ErrStreamAlreadyOpened
	}
	carrier := b.stream.Carrier()
	if carrier != "websocket" && carrier != "tcp" {
		return nil, fmt.Errorf("stream carrier must be websocket or tcp, got %q", carrier)
	}
	id := b.correlator.Take()
	call := CallFrame(
		id,
		b.request.Key,
		b.request.Method,
		b.request.Path,
		b.request.Query,
		b.request.Body,
	)
	session, err := b.stream.Open(ctx, call)
	if err != nil {
		return nil, fmt.Errorf("%s: opening stream failed: %w", b.request.Key, err)
	}
	return &RpcStreamClient[T]{
		session: session,
		id:      id,
		key:     b.request.Key,
		carrier: carrier,
		decode:  b.decode,
	}, nil
}

type FramedStreamTransport struct {
	stream     FramedStream
	correlator *Correlator
}

func NewFramedStreamTransport(stream FramedStream, idPrefix string) *FramedStreamTransport {
	return &FramedStreamTransport{
		stream:     stream,
		correlator: NewCorrelator(idPrefix),
	}
}

func PrepareStream[T any](
	transport *FramedStreamTransport,
	request StreamRequest,
	decode func(json.RawMessage) (T, error),
) *RpcStreamCallBuilder[T] {
	return NewRpcStreamCallBuilder(transport.stream, transport.correlator, request, decode)
}

File: runtime/dart/lib/src/stream.dart
import 'frame.dart';

abstract interface class FramedStreamSession {
  Stream<Frame> get incoming;
  Future<void> cancel();
}

abstract interface class FramedStream {
  String get carrier;
  Future<FramedStreamSession> open(Frame call);
}

final class RpcStreamException implements Exception {
  const RpcStreamException(this.message, {this.code});

  final String message;
  final String? code;

  @override
  String toString() => code == null
      ? 'RpcStreamException: $message'
      : 'RpcStreamException($code): $message';
}

final class RpcStreamContext {
  const RpcStreamContext({
    required this.id,
    required this.key,
    required this.carrier,
    required this.ended,
    required this.cancelled,
    this.error,
  });

  final String id;
  final String key;
  final String carrier;
  final bool ended;
  final bool cancelled;
  final Object? error;
}

final class StreamRequest {
  const StreamRequest({
    required this.key,
    required this.method,
    required this.path,
    this.query = const [],
    this.body,
  });

  final String key;
  final String method;
  final String path;
  final List<QueryPair> query;
  final PresentBody? body;
}

final class RpcStreamClient<T> {
  RpcStreamClient._({
    required FramedStreamSession session,
    required String id,
    required String key,
    required String carrier,
    required T Function(Object?) decoder,
  })  : _session = session,
        _id = id,
        _key = key,
        _carrier = carrier,
        _decoder = decoder;

  final FramedStreamSession _session;
  final String _id;
  final String _key;
  final String _carrier;
  final T Function(Object?) _decoder;

  bool _ended = false;
  bool _cancelled = false;
  Object? _error;

  RpcStreamContext get context => RpcStreamContext(
        id: _id,
        key: _key,
        carrier: _carrier,
        ended: _ended,
        cancelled: _cancelled,
        error: _error,
      );

  Future<void> cancel() async {
    if (_ended || _cancelled) return;
    try {
      await _session.cancel();
      _cancelled = true;
    } on Object catch (error) {
      final wrapped = RpcStreamException('$_key: stream cancellation failed: $error');
      _error = wrapped;
      throw wrapped;
    }
  }

  Stream<T> get values async* {
    var sawTerminal = false;
    try {
      await for (final frame in _session.incoming) {
        if (frame.id != _id) {
          throw RpcStreamException(
            'frame for correlation id ${frame.id} arrived on the stream for $_id',
          );
        }
        switch (frame.kind) {
          case FrameKind.data:
            if (!frame.hasBody) {
              throw const RpcStreamException(
                'a stream data frame arrived without a body',
              );
            }
            try {
              yield _decoder(frame.body);
            } on Object catch (error) {
              throw RpcStreamException(
                '$_key: stream data failed typed decoding: $error',
              );
            }
          case FrameKind.end:
            _ended = true;
            sawTerminal = true;
            return;
          case FrameKind.error:
            final remote = RpcStreamException(
              frame.message ?? 'remote ${frame.code ?? "unknown"}',
              code: frame.code,
            );
            _error = remote;
            sawTerminal = true;
            throw remote;
          case FrameKind.cancel:
            _cancelled = true;
            sawTerminal = true;
            return;
          case FrameKind.call:
            throw const RpcStreamException(
              'a call frame cannot arrive inside its response stream',
            );
        }
      }
      if (!sawTerminal && !_ended && !_cancelled) {
        throw const RpcStreamException(
          'stream transport ended without an end or cancel frame',
        );
      }
    } on Object catch (error) {
      _error = error;
      rethrow;
    }
  }
}

final class RpcStreamCallBuilder<T> {
  RpcStreamCallBuilder._({
    required FramedStream stream,
    required Correlator correlator,
    required StreamRequest request,
    required T Function(Object?) decoder,
  })  : _stream = stream,
        _correlator = correlator,
        _request = request,
        _decoder = decoder;

  final FramedStream _stream;
  final Correlator _correlator;
  final StreamRequest _request;
  final T Function(Object?) _decoder;
  bool _opened = false;

  /// Sole stream I/O boundary. Constructing/configuring this builder performs no I/O.
  Future<RpcStreamClient<T>> stream() async {
    if (_opened) {
      throw StateError('an RPC stream call builder can only be opened once');
    }
    _opened = true;
    if (_stream.carrier != 'websocket' && _stream.carrier != 'tcp') {
      throw RpcStreamException(
        'stream carrier must be websocket or tcp, got ${_stream.carrier}',
      );
    }

    final id = _correlator.take();
    final call = Frame.call(
      id: id,
      key: _request.key,
      method: _request.method,
      path: _request.path,
      query: _request.query,
      body: _request.body,
    );
    late final FramedStreamSession session;
    try {
      session = await _stream.open(call);
    } on Object catch (error) {
      throw RpcStreamException('${_request.key}: opening stream failed: $error');
    }
    return RpcStreamClient<T>._(
      session: session,
      id: id,
      key: _request.key,
      carrier: _stream.carrier,
      decoder: _decoder,
    );
  }
}

final class FramedStreamTransport {
  FramedStreamTransport(this._stream, [String idPrefix = ''])
      : _correlator = Correlator(idPrefix);

  final FramedStream _stream;
  final Correlator _correlator;

  RpcStreamCallBuilder<T> prepare<T>(
    StreamRequest request,
    T Function(Object?) decoder,
  ) =>
      RpcStreamCallBuilder<T>._(
        stream: _stream,
        correlator: _correlator,
        request: request,
        decoder: decoder,
      );
}
